crates/sui-graphql-client/src/faucet.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::time::Duration;
+
+use anyhow::bail;
+use reqwest::{StatusCode, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use sui_types::types::{Address, ObjectId, TransactionDigest};
+
+use crate::{DEVNET_HOST, LOCAL_HOST};
+
+const FAUCET_DEVNET_HOST: &str = "https://faucet.devnet.sui.io/v1/gas";
+const FAUCET_TESTNET_HOST: &str = "https://faucet.testnet.sui.io/v1/gas";
+const FAUCET_LOCAL_HOST: &str = "http://localhost:9123/v1/gas";
+
+const FAUCET_REQUEST_TIMEOUT: Duration = Duration::from_secs(120);
+const FAUCET_POLL_INTERVAL: Duration = Duration::from_secs(2);
+
+pub struct FaucetClient {
+    faucet_url: Url,
+    inner: reqwest::Client,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BatchFaucetResponse {
+    // This string is the Uuid for the req
+    pub task: Option<String>,
+    pub error: Option<String>,
+}
+
+#[derive(serde::Deserialize)]
+pub struct FaucetResponse {
+    task: Option<String>,
+    error: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct BatchStatusFaucetResponse {
+    pub status: Option<BatchSendStatus>,
+    pub error: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum BatchSendStatusType {
+    Inprogress,
+    Succeeded,
+    Discarded,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BatchSendStatus {
+    pub status: BatchSendStatusType,
+    pub transferred_gas_objects: Option<FaucetReceipt>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct FaucetReceipt {
+    pub sent: Vec<CoinInfo>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BatchFaucetReceipt {
+    pub task: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct CoinInfo {
+    pub amount: u64,
+    pub id: ObjectId,
+    pub transfer_tx_digest: TransactionDigest,
+}
+
+impl FaucetClient {
+    /// Construct a new `FaucetClient` with the given `reqwest::Client`. Defaults to Testnet
+    /// faucet.
+    pub(crate) fn new(inner: reqwest::Client, rpc: &Url) -> Self {
```

Comment:
Lets change this to only take a url, the client can be created inside.

Reply:
so not reuse the reqwest::Client we already have? What's the reasoning for having multiple Clients?

Reply:
Maybe in this case it's not that performance critical, but I thought reusing the Client is ok as it is already behind an Arc.

Reply:
Ok, now I understood what you were referring to. Fixed.